src/main/java/org/jvnet/hudson/test/RestartableJenkinsRule.java

```diff
@@ -76,6 +83,63 @@ public void evaluate() throws Throwable {
         });
     }
 
+    /** Approach adapted from https://stackoverflow.com/questions/6214703/copy-entire-directory-contents-to-another-directory */
+    static class CopyFileVisitor extends SimpleFileVisitor<Path> {
+        private final Path targetPath;
+        private Path sourcePath = null;
+        public CopyFileVisitor(Path targetPath) {
+            this.targetPath = targetPath;
+        }
+
+        @Override
+        public FileVisitResult preVisitDirectory(final Path dir,
+                                                 final BasicFileAttributes attrs) throws IOException {
+            if (sourcePath == null) {
+                sourcePath = dir;
+            } else {
+                Files.createDirectories(targetPath.resolve(sourcePath
+                        .relativize(dir)));
+            }
+            return FileVisitResult.CONTINUE;
+        }
+
+        @Override
+        public FileVisitResult visitFile(final Path file,
+                                         final BasicFileAttributes attrs) throws IOException {
+            if (!Files.isSymbolicLink(file)) {
+                // Needed because Jenkins includes invalid lastSuccessful symlinks and otherwise we get a NoSuchFileException
+                Files.copy(file,
+                        targetPath.resolve(sourcePath.relativize(file)));
+            } else if (Files.isSymbolicLink(file) && Files.exists(Files.readSymbolicLink(file))) {
+                Files.copy(file,
+                        targetPath.resolve(sourcePath.relativize(file)));
```

Comment:
Unless I am missing something, these two branches do exactly the same thing.

Reply:
@jglick It's subtle -- think about what happens if you have a symbolic link to something nonexistent. ;)  This is solving a specific bug that occurs with Jenkins.